pkg/streamingpromql/operators/functions/sort_by_label.go

```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package functions
+
+import (
+	"context"
+	"sort"
+
+	"github.com/facette/natsort"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/limiter"
+)
+
+type SortByLabel struct {
+	sortBase
+
+	labels []string
+}
+
+var _ types.InstantVectorOperator = &SortByLabel{}
+
+func NewSortByLabel(
+	inner types.InstantVectorOperator,
+	labels []string,
+	descending bool,
+	memoryConsumptionTracker *limiter.MemoryConsumptionTracker,
+	expressionPosition posrange.PositionRange,
+
+) *SortByLabel {
+	return &SortByLabel{
+		sortBase: newSortBase(inner, descending, memoryConsumptionTracker, expressionPosition),
+		labels:   labels,
+	}
+}
+
+func (s *SortByLabel) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	allSeries, err := s.inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	s.allData = make([]types.InstantVectorSeriesData, len(allSeries))
+	for idx := range allSeries {
+		d, err := s.inner.NextSeries(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		s.allData[idx] = d
+	}
```

Comment:
I don't think it's necessary to load all the data upfront for `sort_by_label` / `sort_by_label_desc` given we're sorting only by the series labels.

I'm imagining we'd do something along the lines of:

1. load all input series' metadata
2. sort the list of series based on their labels, and keep track of which output series corresponds to which output series (eg. the series that sorts first and is returned first is the 3rd input series, the second output series is the 5th input series etc.), and retain this list for use in `NextSeries`
3. use a `InstantVectorOperatorBuffer` to handle buffering series as we need them in `NextSeries`

This might reduce the value of introducing `sortBase` given there will be far less shared code.